### DISCUSSION TITLE
Make a addon that runs after FederatedExtensionsAddon

### DISCUSSION THREAD

**Discussion by martinRenou** [jupyterlite/jupyterlite#647] [Q&A] [answered]
I am making a custom addon which needs to run after the core `FederatedExtensionsAddon`.

The following part of the doc is unclear to me:

> The App-level tasks already have `doit.create_after` configured based on their [hook
parent][hook-parent], which means a Task can _confidently_ rely on files from its
parents (by any other addons) already existing.

I guess I'd like to have my addon to have the `FederatedExtensionsAddon` as parent? Is this possible?

I also saw there was the `file_dep` property on tasks, could I rely on this?

**Comment by bollwyvl** [comment, marked as answer]
There are few topologies: 
- hook-to-hook is a DAG
  - the `pre_` is linked to the `post_` of the one before it 
- hook-to-sub-hook is a strict ordering of `pre_`, _nil_ and `post_`
- sub-hook-to-addon is one-to-many, with otherwise indeterminate ordering

So the `federated_extensions` addon fires during [three hook phases](https://github.com/jupyterlite/jupyterlite/blob/main/py/jupyterlite/src/jupyterlite/addons/federated_extensions.py#L28), namely ` "post_init", "pre_build", "post_build"`, and then have their own subtasks within that.

So to go after `federated_extensions` has done everything it is going to do during `build`, one would need to make a `task_dep = ["post_build:federated_extensions:patch"]`.

> `file_dep` property on tasks

`file_dep` is _usually_ better, but in this case the `post_build:federated_extensions:patch` works against `jupyter-lite.json`, which is a pretty contentious file... likely we could model it better by having all the addons that change those key files instead generate JSON patch files, and then apply those.

**Reply by martinRenou** [reply]
Thanks for your answer!
